### DISCUSSION TITLE
Please clarify how transitive dependencies resolve in a workspace

### DISCUSSION THREAD

**Discussion by Aghassi** [pnpm/pnpm#7392] [Q&A]
I'm running into a very frustrating issue and only have found one solution so I want to confirm the "right" approach.

I have a monorepo that has the following structure

```
package.json
|
| - project-a/
     | - package.json
| - project-b/
     | - package.json
```

Currently, `project-a` has `@types/react` pinned to `16.14.21` and has dev dependencies that transitively rely on `@types/react@*`
When `project-b` adds `@types/react` version `18.2.42`, it rewrites the transitive dependencies in `project-a` to use `@types/react` `18.2.42` instead of leaving them pinned to `16.14.21` like they were before.

My question is: how do I prevent this behavior? The best I have found is to list each package that has a transitive dependency of `@types/react@*` in `.pnpmfile.cjs` such that it pins to the version I want.

What the best way to solve this problem because it's unnerving to have transitive dependencies shifting versions like this when two projects don't rely on each other in anyway.

**Comment by thynson** [comment]
From your description, I assume that these transitive dependencies must share same instance of `@types/react` with their consumer (`project-a` and `project-b` here), these packages should declare `@types/react` as a peer dependencies.
If these packages are under your control, fix them by putting `"@types/react": "*"` in `peerDependencies` instead of `dependencies`.
If they are not authored and published by you, use [`pnpm patch`](https://pnpm.io/cli/patch) to work around.

**Reply by Aghassi** [reply]
Yeah they use them in `dependencies`. If we patch them though, that patch gets applied after resolution and lockfile generation though right? Or am I misunderstanding how the files get laid out and resolved. I can also use `pnpmfile.cjs` to do the same thing by rewriting the package.json on the fly. 

I'm having trouble more broadly knowing when to use `packageExtension`, `.pnpmfile.cjs`, or `pnpm patch`
